sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -142,7 +147,20 @@ def split(
         stop_position=stop_position)
 
   def estimate_size(self):
-    return self._get_concat_source().estimate_size()
+    file_names = [f for f in fileio.ChannelFactory.glob(self._pattern)]
+    if (len(file_names) <=
+        FileBasedSource.MAX_NUMBER_OF_FILES_FOR_AN_EXACT_STAT):
+      return sum(self._estimate_sizes_in_parallel(file_names))
+    else:
+      # Estimating size of a random sample.
+      sample_size = max(FileBasedSource.MAX_NUMBER_OF_FILES_FOR_AN_EXACT_STAT,
```

Comment:
This assumes that all the files have about the same size. It might be better to sample until we have reasonable confidence in the file sizes by looking at the distribution of file sizes sampled so far. (You can add a TODO if you want to defer this).